src/avram/errors.cr

```diff
@@ -14,40 +14,53 @@ module Avram
   # Raised when trying to access a record that was not preloaded and lazy load
   # is disabled.
   class LazyLoadError < AvramError
-    def initialize(model : String, association : String)
+    getter model
+    getter association
+
+    def initialize(@model : String, @association : String)
       super "#{association} for #{model} must be preloaded with 'preload_#{association}'"
     end
   end
 
   # Raised when Avram cannot find a record by given id
   class RecordNotFoundError < AvramError
-    def initialize(model : TableName, id : String)
+    getter model
+    getter id = nil
+    getter query = nil
+
+    def initialize(@model : TableName, @id : String)
       super "Could not find #{model} with id of #{id}"
     end
 
-    def initialize(model : TableName, query : Symbol)
+    def initialize(@model : TableName, @query : Symbol)
       super "Could not find #{query} record in #{model}"
     end
   end
 
   class MissingRequiredAssociationError < AvramError
-    def initialize(model : Avram::Model.class, association : Avram::Model.class)
+    getter model
+    getter association
+
+    def initialize(@model : Avram::Model.class, @association : Avram::Model.class)
       super "Expected #{model} to have an association with #{association} but one was not found."
     end
   end
 
   # Raised when a validation is expecting an impossible constraint
   class ImpossibleValidation < AvramError
-    def initialize(attribute : Symbol, message = "an impossible validation")
+    getter attribute
+
+    def initialize(@attribute : Symbol, message = "an impossible validation")
       super "Validation for #{attribute} can never satisfy #{message}"
     end
   end
 
   # Raised when using the create!, update!, or delete! methods on an operation when it does not have the proper attributes
-  class InvalidOperationError < AvramError
+  class InvalidOperationError(T) < AvramError
```

Comment:
I think the Generic Type can be left off this, then the constructor can use `forall T` instead.